File: frontend/types/bullet.ts
export interface ResumeBulletItem {
  section: "experience" | "projects";
  entry_index: number;
  bullet_index: number;
  entry_title: string | null;
  organization: string | null;
  text: string;
}

export interface ImproveBulletRequest {
  bullet_text: string;
  resume_id?: string;
  resume_version_id?: string | null;
  target_role?: string | null;
  regenerate?: boolean;
  previous_improved_text?: string | null;
}

export interface ImproveBulletResponse {
  original_text: string;
  improved_text: string;
  changes_summary: string;
  metric_placeholder_used: boolean;
  suggested_metric_prompt: string | null;
  regenerate: boolean;
  prompt_version: string;
}

export interface ReplaceBulletRequest {
  resume_id: string;
  resume_version_id?: string | null;
  section: "experience" | "projects";
  entry_index: number;
  bullet_index: number;
  improved_text: string;
}

export interface ReplaceBulletResponse {
  resume_id: string;
  section: string;
  entry_index: number;
  bullet_index: number;
  updated_text: string;
  message: string;
}

export type BulletImproveStage = "idle" | "improving" | "done" | "error";

File: backend/app/schemas/bullet.py
"""Bullet improvement API schemas."""

from __future__ import annotations

import uuid
from typing import Literal

from pydantic import Field

from app.schemas.common import BaseSchema


class ImproveBulletRequest(BaseSchema):
    bullet_text: str = Field(min_length=1)
    resume_id: uuid.UUID | None = None
    resume_version_id: uuid.UUID | None = None
    target_role: str | None = Field(default=None, max_length=255)
    regenerate: bool = False
    previous_improved_text: str | None = None


class ImproveBulletResponse(BaseSchema):
    original_text: str
    improved_text: str
    changes_summary: str
    metric_placeholder_used: bool = False
    suggested_metric_prompt: str | None = None
    regenerate: bool = False
    prompt_version: str


class ReplaceBulletRequest(BaseSchema):
    resume_id: uuid.UUID
    resume_version_id: uuid.UUID | None = None
    section: Literal["experience", "projects"]
    entry_index: int = Field(ge=0)
    bullet_index: int = Field(ge=0)
    improved_text: str = Field(min_length=1)


class ReplaceBulletResponse(BaseSchema):
    resume_id: uuid.UUID
    section: str
    entry_index: int
    bullet_index: int
    updated_text: str
    message: str


class ResumeBulletItem(BaseSchema):
    section: str
    entry_index: int
    bullet_index: int
    entry_title: str | None = None
    organization: str | None = None
    text: str
